### PR TITLE
assorted fixes for powerpc-apple-darwin9

### PR DESCRIPTION
**Pull request by grobian**
system info:

% gcc -dumpmachine
powerpc-apple-darwin9
% gcc -dumpversion
10.1.0
% ld -v
@(#)PROGRAM:ld  PROJECT:ld64-97.17 (Gentoo binutils-apple-3.2.6)
Library search paths:
        /Volumes/Scratch/Gentoo/Leopard/usr/lib
        /Volumes/Scratch/Gentoo/Leopard/lib
        /usr/lib
        /usr/local/lib
Framework search paths:
        /Library/Frameworks/
        /System/Library/Frameworks/
ld: warning: -arch not specified
ld: no object files specified for inferred architecture ppc

### CODE REVIEW

**Comment by cooljeanius** [comment]
Um... this seems like kind of a lot? Do they all really need to be combined into a single PR?

**Comment by grobian** [comment]
hmmyeah, not sure what went wrong here, these are now fixes for Darwin in general.

**Comment by cooljeanius** [comment]
also it looks like this includes PR #8 as well? It'd probably be better to get that one merged separately first...

**Comment by iains** [comment]
hmm I have no idea how I missed this PR (somehow I got no email) only spotted because of the #7 patch.

(a) I guess this is still relevant?
(b) would it be possible to split it into pieces? [GH does not, AFAIU, allow me to ack individual commits while not merging the whole branch]..  Ideally a branch should address a single topic.

**Comment by grobian** [comment]
Ok, I should probably revive this then and make it more consumable.  Of course you can cherry pick commits if you see something useful.  I was under the impression there was no interest in it, and I think I pushed additional fixes to the same branch or something.  I'll look into cleaning this up.

**Comment by iains** [comment]
sadly, this is not $dayjob - it's not lack of interest, just very limited time :(

Actually, I am willing to cherry-pick individual commits - if that suits you.

For information:  I have a much more compatible xtools branch for OSX-10.4 - 10.8 (which includes the ability to run the ld64 tests) - it is buildable starting from Apple's gcc-4.2.1.  This will become xtools-0.7.0 and will be recommended for all archs in that OS range)

I am hoping to get some cycles on xtools in November.

**Comment by grobian** [comment]
Same here, and I frankly forgot about this.  Certainly looking forward to using a newer xtools (0.7.0 being newer than 2.2.4?).  It's the default linker on Gentoo for macOS, except arm64 where it doesn't work.

**Comment by iains** [comment]
> Same here, and I frankly forgot about this. Certainly looking forward to using a newer xtools (0.7.0 being newer than 2.2.4?).

The idea is to try to move to building whole toolchains that are matched to the OS versions they are trying to support - so that:

 * 0.7.0 would be appropriate for 10.4-10.8 
 * 2.2.4-some-release-version would be appropriate for 10.9 - 1011
 * something based on Xcode 10.3 sources for 10.12-10.14
 * etc.

"One-size-fits-all" is an attractive idea - but it gets increasingly complex to maintain.

e.g. - One issue is that 32b support got ripped out - so that if we try to support powerpc on "latest available" Xcode sources we are not only adding back powerpc but also the 32b stuff.  Not impossible (it's all "just code") but time taken could be significant.

edit: A secondary motivation is to pick the most modern toolchain components that we can realistically bootstrap on the target using only the available Xcode version (so, for example, on 10.5 we need to start from gcc-4.2.1 + ld64-85.2.1 + dwarfutils dsymutil and use that to bootstrap a capable C++11/4 toolchain).  Moving much beyond ld64-236 really means using a C++11 compiler (hence the recommendation to use 0-7-0 for the platform versions that do not have a usable C++11 compiler installed).

> It's the default linker on Gentoo for macOS, except arm64 where it doesn't work.

Likewise, we have to make changes to the latest published - (Xcode 12.0 sources, currently) and then add that to the arm64 braches (also here) - although we are trying to upstream the arm64 support.

File: cctools/PB.project
Comment (cooljeanius) [inline cctools/PB.project:1]: OK, so, yes, this file is old and obsolete, but would it really harm anything just to leave it?

File: cctools/ar/ar.c [hunk 1]
```diff
@@ -123,12 +123,12 @@ main(argc, argv)
...
 
 	if (argc < 3) {
-	   if(strcmp(argv[1], "--version") == 0){
+	   if(argc >= 2 && strcmp(argv[1], "--version") == 0){
 		/* Implement a gnu-style --version to be friendly to GCC.  */
 		fprintf(stdout, "xtools-%s ar %s\nBased on Apple Inc. %s\n",
 		        xtools_version, package_version, apple_version);
 		exit(0);
-	   } else if(strcmp(argv[1], "--help") == 0){
+	   } else if(argc >= 2 && strcmp(argv[1], "--help") == 0){
```
Comment (cooljeanius) [inline cctools/ar/ar.c:131]: These changes were already included in #8 which has been merged; I wonder why they're still showing up as part of the diff here?

File: cctools/ar/archive.c [hunk 2]
```diff
@@ -343,7 +346,8 @@ put_arobj(cfp, sb)
 			    sb->st_mode, sb->st_size, ARFMAG);
 			lname = 0;
 		} else if (lname > sizeof(hdr->ar_name) || strchr(name, ' '))
-			(void)sprintf(hb, HDR1, AR_EFMT1, (lname + 3) & ~3,
+			(void)sprintf(hb, HDR1, AR_EFMT1,
```
Comment (cooljeanius) [inline cctools/ar/archive.c:349]: I wonder if it'd be worthwhile to change `sprintf()` to `snprintf()` while you're here?

File: cctools/as/arm.c [hunk 3]
```diff
@@ -1448,7 +1448,7 @@ parse_reg_list (char ** strp)
 	      cur_reg = reg;
 	    }
 	  while (skip_past_comma (&str) != FAIL
-		 || (in_range = 1, *str++ == '-'));
+		 || ((void)(in_range = 1), *str++ == '-'));
```
Comment (cooljeanius) [inline cctools/as/arm.c:1451]: if this is just to silence `-Wcomma`, I generally think it's better to just ignore that warning, or to apply some other method of solving it, instead of applying its suggestion to add a `(void)` cast. But I guess that's a matter of opinion.

File: cctools/as/arm.c [hunk 4]
```diff
@@ -2175,7 +2175,7 @@ create_neon_reg_alias (char *newname, char *p)
         }
     }
 
-  namelen = nameend - newname;
+  namelen = (int)(nameend - newname);
```
Comment (cooljeanius) [inline cctools/as/arm.c:2178]: instead of casting to `int`, I think it might be better to just change the type of `namelen` where it's originally declared. It looks like `ptrdiff_t` might be the appropriate type here? Or maybe `size_t`, based on how it's used later...

File: cctools/as/arm.c [hunk 5]
```diff
@@ -16956,10 +16966,12 @@ md_apply_fix (fixS *	fixP,
     case BFD_RELOC_ARM_OFFSET_IMM:
 #ifdef NOTYET
       if (!fixP->fx_done && seg->use_rela_p)
+	  value = 0;
 #else
-      if (!fixP->fx_done && 0)
+// The goggles do nothing!
```
Comment (cooljeanius) [inline cctools/as/arm.c:16971]: instead of commenting this out, maybe just change the preprocessor conditional that it's in?

File: cctools/as/dwarf2dbg.c [hunk 6]
```diff
@@ -1840,8 +1842,9 @@ struct frchain * abbrev_section)
     {
       if (DWARF2_FORMAT () == dwarf2_format_32bit)
 	out_abbrev (DW_AT_ranges, DW_FORM_data4);
-      else
-	out_abbrev (DW_AT_ranges, DW_FORM_data8);
+// The goggles do nothing
```
Comment (cooljeanius) [inline cctools/as/dwarf2dbg.c:1845]: Instead of commenting this out, maybe just change the conditional?

File: cctools/as/i386.c [hunk 7]
```diff
@@ -4804,7 +4805,7 @@ i386_operand (operand_string)
      char *operand_string;
 {
   const reg_entry *r;
-  char *end_op;
+    char *end_op = NULL;
```
Comment (cooljeanius) [inline cctools/as/i386.c:4808]: unnecessary whitespace added as indentation here?
```suggestion
  char *end_op = NULL;
```

File: cctools/as/i386.c [hunk 8]
```diff
@@ -5408,6 +5409,8 @@ int md_long_jump_size = 5;
 /* Size of relocation record.  */
 const int md_reloc_size = 8;
 
+void md_create_short_jump (char *ptr, addressT from_addr, addressT to_addr, fragS *, symbolS *to_symbol);
```
Comment (cooljeanius) [inline cctools/as/i386.c:5412]: So, what I'd do here would be to move the newly-added prototypes to `i386.h`, since they're extern, and then update the function definitions here so that they're no longer old-style.

File: cctools/as/write_object.c [hunk 9]
```diff
@@ -1115,6 +1119,9 @@ uint32_t debug_section)
     uint32_t left21, right14;
 #endif
 
+/* the pragmas that follow silence a noisy clang warning  */
```
Comment (cooljeanius) [inline cctools/as/write_object.c:1122]: maybe put them in an `#ifdef __clang__` preprocessor conditional, then?

File: cctools/include/stuff/arch.h [hunk 10]
```diff
@@ -97,6 +97,14 @@ __private_extern__ const char *get_arch_name_from_types(
     cpu_type_t cputype,
     cpu_subtype_t cpusubtype);
 
+/*
+ * get_arch_name_if_known() returns the name of the architecture for the
+ * specified cputype and cpusubtype if known.  If unknown it returns NULL.
+ */
+__private_extern__ const char *get_arch_name_if_known(
```
Comment (cooljeanius) [inline cctools/include/stuff/arch.h:104]: Isn't `__private_extern__` deprecated? See the following GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=106672

File: cctools/include/stuff/bytesex.h [hunk 11]
```diff
@@ -317,6 +317,10 @@ __private_extern__ void swap_arm_thread_state64_t(
     arm_thread_state64_t *cpu,
     enum byte_sex target_byte_sex);
 
+__private_extern__ void swap_arm_exception_state64_t(
```
Comment (cooljeanius) [inline cctools/include/stuff/bytesex.h:320]: Again with the `__private_extern__`s in this file, too.

File: cctools/ld/layout.c [hunk 12]
```diff
@@ -2553,3 +2557,66 @@ print_thread_info(void)
...
+ * architectures have come and gone and come back, you can't include all of
+ * these headers in one source and some of the constants covered the whole
+ * address space the common value of 64meg was chosen.
+ */
+static
+uint32_t
+get_stack_size_from_flag(
+			 const struct arch_flag *flag)
+{
+#ifdef __MWERKS__
```
Comment (cooljeanius) [inline cctools/ld/layout.c:2575]: New code with an `#ifdef __MWERKS__`? Does anyone still even have a Metrowerks compiler to test with? Also, this code looks like it's just attempting to avoid a warning about an unused parameter, and I thought that Metrowerks also supported GCC's `#pragma unused` for that purpose?

File: cctools/libmacho/Makefile [hunk 13]
```diff
@@ -39,6 +39,10 @@ CFLAGS = $(OFLAG) -g -Wall $(LEGACY) \
 	 -I$(SRCROOT) -I$(SRCROOT)/../include -I$(OFILE_DIR) $(SDK)
 PG = -pg
 
+ifeq "$(RC_MACOS)" "YES"
+  DYNAMICLIB_FLAGS = -Wl,-iosmac_version_min,12.0
```
Comment (cooljeanius) [inline cctools/libmacho/Makefile:43]: The variable being checked says MACOS, but the flag being used is an iOS one? Is that correct?

File: cctools/libstuff/guess_short_name.c [hunk 14]
```diff
@@ -59,6 +59,14 @@ static char *look_back_for_slash(
  * If the name of the dynamic library is none of the forms above then NULL is
  * returned.
  */
+/*
+ * MDT 20190119 rdar://12400897
```
Comment (cooljeanius) [inline cctools/libstuff/guess_short_name.c:63]: Has this one been posted to [OpenRadar](https://openradar.appspot.com/page/1)? I looked for it there, but couldn't find it...

File: cctools/libstuff/ofile.c [hunk 15]
```diff
@@ -919,6 +920,9 @@ enum bool archives_with_fat_objects)
     uint64_t offset;
     uint32_t align;
 
+	/* silence clang warning */
```
Comment (cooljeanius) [inline cctools/libstuff/ofile.c:923]: which flag does it come from, specifically?
(likewise, wherever else a comment mentions silencing a warning, please update those comments to say which one, as well)

**Comment by grobian** [comment]
I need to sort this out, but my PowerMac is dying (it seems) and attempt to get Intel-based fallback hasn't been much of a success either.

I think most of these diffs are not the once I introduced, because for some reason I messed up this PR.  So I'm going to close it because it's obviously gibberish at this point, and I can't tell what should be good or bad or necessary.  All I know is it works fine on x86_64 on Ventura still with GCC.
